### ISSUE TITLE
use pytest

### ISSUE THREAD

**Issue by sils1297** [coala-analyzer/coala#875]
http://pytest.org/latest/goodpractises.html#integrating-with-distutils-python-setup-py-test

should be easy and cool :)

**Comment by sils1297** [comment]
maybe we want tox too, sounds like it could even simplify our CI scripting.

**Comment by sils1297** [comment]
So I'm not sure if we really want that.

1. I like our output better
2. Our test framework seems just faster
3. Our test framework just executes our test file and the other framework executes the Unittest objects - because of that it doesn't execute the module code and that leads to failing tests.

I started migrating out our test framework to cotest, not sure though if we shouldn't just make our tests independent on module code and use pytest. Any opinions?

**Comment by AbdealiJK** [comment]
Everything you said has plugins in pytest.

There must be something for formatting
Three a plugin to run tests in parallel
There must be a way to skip tests if needed

I think it may be wiser to make our own pytest plugins if there are any issues with it in general


-----Original Message-----

**Comment by sils1297** [comment]
@AbdealiJK following tests fail when executed with pytest:

http://pastebin.com/F1NWHCV8

**Comment by sils1297** [comment]
Marking wontfix, for now our framework works and we need to focus.

**Comment by sils1297** [comment]
I think we'd greatly profit from using doctest which pytest provides so reopening this.

**Comment by Makman2** [comment]
Did you miss to reopen?

**Comment by sils1297** [comment]
@AbdealiJK DBus test is a huge problem because you rely on a custom main routine. Can you help? Currently looking into the rest...

**Comment by AbdealiJK** [comment]
https://pypi.python.org/pypi/pytest-xprocess seems to be the answer to your issue. It spawns a process at the beginning and keeps that process going throughout the testing time. 

Although we want it only for 1 test, this seems the way to go.

-----Original Message-----

**Comment by sils1297** [comment]
@AbdealiJK sure that's the way to go? PyTest also supports test functions, so instead of putting the process spawning into the main routine we could put it into a `test_something` method which invokes another python file (cleaner anyway, one file for the actual test, another file for the coala dbus server)?

**Comment by AbdealiJK** [comment]
I tried that when I first mentioned pytest (as far as I remember). It
didn't work out properly if I remember correctly.
Basically the test_* don't run parallelly on a single core and some other
complications that I don't remember made me not go through that route.

I think it's better to spawn a test dbus server - that way even
DbusDocumentTest.py and other Dbus*Test.py can also test do unit tests
against the server later (right now they don't)

**Comment by sils1297** [comment]
@AbdealiJK do you have any hint about the other failures in pytest? Many are because of the test skipping but some seem to be very strange, like some tests of `is_applicable` of some actions which are really simple.

**Comment by AbdealiJK** [comment]
What exactly do you mean by the failures of pytest ? The major issues I
noticed with pytest was :
 - the thing with dbus where we needed an external app running
 - the skip_test functionality needed to be revamped
 - the fact that it was slower than our tests quite a bit
 - And in general lack of flexibility as it's not our code and we had to
follow someone else's docs
Is that what you were looking for ?
The first two took time and we didn't want to spend time on it earlier.
The second two were a trade off between maintaining our own barebones code
and using a full-fledged 3rd part library.

And what did you mean by the is_applicable thing ? What's strange about it ?

**Comment by sils1297** [comment]
FWIW I'm almost done with replacing our skip functionality, @AbdealiJK can you look into DBus?

**Comment by sils1297** [comment]
Ok, that's what I get after https://github.com/coala-analyzer/coala/pull/1183:
http://pastebin.com/ZrMy5D3k

**Comment by sils1297** [comment]
Alright, the DBusTest issue is resolved, think we can make the switch soon

**Comment by AbdealiJK** [comment]
Now a new issue came up. If I do `py.test -k "coalaTest"` the test fails. This happens in `coalaTest` `ConfigurationGatheringTest` and `CliParserTest` - probably because they all use `sys.argv`
